tripy/tripy/constraints.py

```diff
@@ -58,18 +124,104 @@ def get_arg_dtype(arg, func_name, arg_name) -> utils.Result["tripy.dtype"]:
+                else:
+                    all_args[index] = (name, new_args[name])
+
+    # Materialize type variables from tensors.
+    type_vars = {}
+    for name, arg in all_args:
+        if name in constraints:
+            dtype_result = get_arg_dtype(arg, func.__qualname__, name)
+            if dtype_result:
+                type_vars[constraints[name]] = dtype_result.value
+
+    new_args = []
+    new_kwargs = {}
+    for index, (name, arg) in enumerate(all_args):
+
+        def add_arg(arg):
+            if name not in kwargs:
+                new_args.append(arg)
+            else:
+                new_kwargs[name] = arg
+
+        if name in conversion_targets and not isinstance(arg, Tensor):
+            if name in shape_likes:
+                arg = tensor_from_shape_like(arg)
+            else:
+                # Python integers can always be cast to the most restrictive type, which is DimensionSize in Tripy.
+                # DimensionSize can always be cast up to Tensor if needed, but the reverse is not true.
+                # NOTE: We do not use isinstance here because bool is a subclass of int.
+                arg = DimensionSize.create_directly(arg) if type(arg) is int else Tensor.create_directly(arg)
+
+            _add_column_info(
+                arg,
+                index,
+                name in kwargs,
+                len(args),
+                func.__name__,
+                [name for name, _ in all_args],
+            )
+
+            dtype = None
+            if name in constraints and constraints[name] in type_vars:
+                dtype = type_vars[constraints[name]]
+
+            if dtype is not None:
+                # Refuse to do unsafe casts like float -> int.
+                # NOTE: We do this check all the way down here so we have better stack information for `arg`.
+                UNSAFE_CASTS = {floating: integer, integer: tp_bool}
+                if any(issubclass(arg.dtype, frm) and issubclass(dtype, to) for frm, to in UNSAFE_CASTS.items()):
+                    raise_error(
+                        f"Refusing to automatically cast '{arg.dtype}' argument to '{dtype}'.",
+                        [
+                            f"Hint: You may want to manually cast other operands in this expression to compatible types.\n",
+                            "Note: argument was: ",
+                            arg,
+                        ],
+                    )
+
+                original_stack_info = arg.stack_info
+                arg = cast(arg, dtype=dtype)
+                arg.stack_info = original_stack_info
+
+        add_arg(arg)
+
+    return new_args, new_kwargs
+
+
 def dtypes(
     constraints: Dict[str, str] = {},
     variables: Dict[str, List[str]] = {},
     exceptions: List[Dict[str, str]] = [],
     aliases: List[str] = [],
+    convert_tensor_and_shape_likes: bool = False,
+    conversion_targets: Set[str] = None,
```

Comment:
Does it make sense to unify these? We can have:
```py
convert_to_tensors: Union[bool, Set[str]]
```
where `convert_to_tensors=True` means that we should automatically choose what to convert while `convert_to_tensors=[x, y, z]` means we should convert only `x`, `y`, and `z`.